Mathlib/FieldTheory/Differential/Liouville.lean

```diff
@@ -0,0 +1,175 @@
+    IsLiouville F B where
+  is_liouville (a : F) (ι : Type) [Fintype ι] (c : ι → F) (hc : ∀ x, (c x)′ = 0)
+      (u : ι → B) (v : B) (h : a = ∑ x, c x * logDeriv (u x) + v′) := by
+    apply inst.is_liouville a ι c hc (B.val ∘ u) (B.val v)
+    dsimp only [coe_val, Function.comp_apply]
+    conv =>
+      rhs
+      congr
+      · rhs
+        intro x
+        rhs
+        apply logDeriv_algebraMap (u x)
+      · apply (deriv_algebraMap v)
+    simp_rw [IsScalarTower.algebraMap_apply F B K]
+    norm_cast
+
+
+lemma IsLiouville.equiv {K' : Type*} [Field K'] [Differential K'] [Algebra F K']
+    [DifferentialAlgebra F K'] [Algebra.IsAlgebraic F K']
+    [inst : IsLiouville F K] (e : K ≃ₐ[F] K') : IsLiouville F K' where
+  is_liouville (a : F) (ι : Type) [Fintype ι] (c : ι → F) (hc : ∀ x, (c x)′ = 0)
+      (u : ι → K') (v : K') (h : a = ∑ x, c x * logDeriv (u x) + v′) := by
+    apply inst.is_liouville a ι c hc (e.symm ∘ u) (e.symm v)
+    apply_fun e.symm at h
+    simpa [AlgEquiv.commutes, map_add, map_sum, map_mul, logDeriv, algEquiv_deriv'] using h
+
+private local instance liouville_of_finiteDimensional_galois [FiniteDimensional F K]
+    [IsGalois F K] : IsLiouville F K where
+  is_liouville (a : F) (ι : Type) [Fintype ι] (c : ι → F) (hc : ∀ x, (c x)′ = 0)
+      (u : ι → K) (v : K) (h : a = ∑ x, c x * logDeriv (u x) + v′) := by
+    haveI : CharZero K := charZero_of_injective_algebraMap
+      (NoZeroSMulDivisors.algebraMap_injective F K)
+    let c' (i : ι) := (c i) / (Fintype.card (K ≃ₐ[F] K))
+    let u'' (i : ι) := ∏ x : (K ≃ₐ[F] K), x (u i)
+    have : ∀ i, u'' i ∈ fixedField (⊤ : Subgroup (K ≃ₐ[F] K)) := by
+      rintro i ⟨e, _⟩
+      change e (u'' i) = u'' i
+      simp only [u'', map_prod]
+      apply Fintype.prod_equiv (Equiv.mulLeft e)
+      simp
+    have ffb := (IsGalois.tfae.out 0 1).mp (inferInstanceAs (IsGalois F K))
+    simp_rw [ffb, IntermediateField.mem_bot, Set.mem_range] at this
+    choose u' hu' using this
+    let v'' := (∑ x : (K ≃ₐ[F] K), x v) / (Fintype.card ((K ≃ₐ[F] K)))
+    have : v'' ∈ fixedField (⊤ : Subgroup (K ≃ₐ[F] K)) := by
+      rintro ⟨e, _⟩
+      change e v'' = v''
+      simp only [v'', map_div₀, map_sum, map_natCast]
+      congr 1
+      apply Fintype.sum_equiv (Equiv.mulLeft e)
+      simp
+    rw [ffb, IntermediateField.mem_bot] at this
+    obtain ⟨v', hv'⟩ := this
+    exists ι, inferInstance, c', ?_, u', v'
+    · intro x
+      simp [c', Derivation.leibniz_div, hc]
+    · apply_fun (algebraMap F K)
+      case inj =>
+        exact NoZeroSMulDivisors.algebraMap_injective F K
+      simp only [map_add, map_sum, map_mul, ← logDeriv_algebraMap, hu', ← deriv_algebraMap, hv']
+      unfold u'' v'' c'
+      clear c' u'' u' hu' v'' v' hv'
+      push_cast
+      rw [Derivation.leibniz_div_const, smul_eq_mul, inv_mul_eq_div]
+      case h => simp
+      simp only [map_sum, div_mul_eq_mul_div]
+      rw [← sum_div, ← add_div]
+      field_simp
+      conv =>
```

Comment:
This part of the proof is illegible. I feel some rewriting lemma is missing, but it is true that I'm not familiar with this use of tactics.

Reply:
I added a comment, does it help?